src/runtime/runtime.ts

```diff
@@ -288,86 +336,357 @@ class Scan {
-      if(prefix[this.n.offset] !== undefined && prefix[this.n.offset] !== input.n) return false;
+      if(prefix[this.n.offset] !== undefined && prefix[this.n.offset] !== input.n) return ApplyInputState.fail;
       prefix[this.n.offset] = input.n;
     }
 
+    return ApplyInputState.pass;
+  }
+
+  propose(index:Index, prefix:ID[], transaction:number, round:number, results:any[]):Proposal {
+    let {e,a,v,n} = this.resolve(prefix);
+    this.proposal.skip = false;
+    let proposal = index.propose(this.proposal, e, a, v, n, transaction, round);
+
+    let ix = 0;
+    for(let field of proposal.forFields) {
+      proposal.forRegisters[ix] = this[field] as Register;
+      ix++;
+    }
+    if(proposal.forFields.length === 0) proposal.skip = true;
+    return proposal;
+  }
+
+  resolveProposal(index:Index, proposal:Proposal, transaction:number, round:number, results:any[]):ID[][] {
+    return index.resolveProposal(proposal);
+  }
+
+  accept(index:Index, prefix:ID[], transaction:number, round:number, solvingFor:Register[]):boolean {
+    // before we start trying to accept, we need to make sure we care about the registers
+    // we are currently solving for
+    let solving = false;
+    for(let register of solvingFor) {
+      if(this.registerLookup[register.offset]) {
+        solving = true;
+        break;
+      }
+    }
+    // if we aren't looking at any of these registers, then we just say we accept
+    if(!solving) return true;
+    let {e,a,v,n} = this.resolve(prefix);
+    let results = index.get(e, a, v, n, transaction, round)
+    return results.length > 0;
+  }
+
+  acceptInput(index:Index, input:Change, prefix:ID[], transaction:number, round:number):boolean {
+    let {e,a,v,n} = this.resolve(prefix);
+    if((e === IGNORE_REG || input.e === e) &&
+       (a === IGNORE_REG || input.a === a) &&
+       (v === IGNORE_REG || input.v === v) &&
+       (n === IGNORE_REG || input.n === n)) {
+      return true;
+    } else  {
+      return this.accept(index, prefix, transaction, round, this.registers);
+    }
+  }
+
+
+  // Scans don't have any inherent setup
+  setup() {
+    if(isRegister(this.e)) this.registers.push(this.e);
+    if(isRegister(this.a)) this.registers.push(this.a);
+    if(isRegister(this.v)) this.registers.push(this.v);
+    if(isRegister(this.n)) this.registers.push(this.n);
+    for(let register of this.registers) {
+      this.registerLookup[register.offset] = true;
+    }
+  }
+
+  getRegisters():Register[] {
+    return this.registers;
+  }
+
+}
+
+//------------------------------------------------------------------------
+// Constraints
+//------------------------------------------------------------------------
+
+enum ApplyInputState {
+  pass,
+  fail,
+  none,
+}
+
+interface Constraint {
+  setup():void;
+  getRegisters():Register[];
+  applyInput(input:Change, prefix:ID[]):ApplyInputState;
+  propose(index:Index, prefix:ID[], transaction:number, round:number, results:any[]):Proposal;
```

Comment:
Is `results` argument necessary for `propose` function? It's not used anywhere yet, will it be?